### PR TITLE
Reject client requests in case the system is not ready

### PR DESCRIPTION
**Pull request by yuliasherman**
When the system starts up the replicas become available within a few minutes and not all together, which causes the system to start a chain of view-changes. To avoid this behavior, SimpleCLient should verify that there are enough live connections to the replicas before starting request handling.

https://jira.eng.vmware.com/browse/BC-2827

### CODE REVIEW

**Comment by teoparvanov** [comment]
Hi Yulia, I suppose with this fix you could also remove the 3 "sleeps" for 5 seconds in the pre-execution tests? For example:
https://github.com/vmware/concord-bft/blob/master/tests/apollo/test_skvbc_preexecution.py#L154

**Comment by yuliasherman** [comment]
> Hi Yulia, I suppose with this fix you could also remove the 3 "sleeps" for 5 seconds in the pre-execution tests? For example:
> https://github.com/vmware/concord-bft/blob/master/tests/apollo/test_skvbc_preexecution.py#L154

Actually, this change is not related to the issue I have with a thread pool, which still requires a fix.

**Comment by teoparvanov** [comment]
Hmm, interesting... So your goal is to avoid the initial view changes if the replicas aren't up and running yet? If so, the following test I recently introduced should fail, right?

https://github.com/vmware/concord-bft/blob/master/tests/apollo/test_skvbc.py#L220

**Comment by yuliasherman** [comment]
You don't expect to have view changes in this situation. But all those tests are using UDP, which is connectionless, while my fix makes sense only for TCP/TLS.

**Comment by teoparvanov** [comment]
> 
> 
> You don't expect to have view changes in this situation. But all those tests are using UDP, which is connectionless, while my fix makes sense only for TCP/TLS.

You are right, thanks for the clarification!

File: bftengine/src/bftengine/SimpleClientImp.cpp [hunk 1]
```diff
@@ -108,6 +111,21 @@ class SimpleClientImp : public SimpleClient, public IReceiver {
   concordlogger::Logger logger_ = concordlogger::Log::getLogger("concord.bft.client");
 };
 
+bool SimpleClientImp::isSystemReady() const {
+  uint16_t connectedReplicasNum = 0;
+  if (communication_->isRunning()) {
+    for (uint16_t rid : replicas_) {
+      if (communication_->getCurrentConnectionStatus(rid) == ConnectionStatus::Connected) connectedReplicasNum++;
```
Comment (andrewjstone) [inline bftengine/src/bftengine/SimpleClientImp.cpp:118]: Unfortunately, I don't think this works. The implementation of getCurrentConnectionStatus in TLS is [incorrect](https://github.com/vmware/concord-bft/blob/master/communication/src/TlsTCPCommunication.cpp#L1287). I discovered this when working on a refactoring a couple months ago. Unfortunately, I also did not have time to finish that (it's about 95%), so the fix is not in. 

You may be able to take my fix although you may need to modify it to fit the exact names in the current code base: 

```C++
ConnectionStatus TlsTCPCommunication::TlsTcpImpl::getCurrentConnectionStatus(const NodeNum id) const {
  std::lock_guard<std::mutex> lock(connectionsGuard_);
  if (connections_.count(id) == 1) {
    return ConnectionStatus::Connected;
  }
  return ConnectionStatus::Disconnected;
}
```

Comment (yuliasherman) [reply bftengine/src/bftengine/SimpleClientImp.cpp:118]: Thanks for the observation, Andrew. I've also noticed that this functionality is not implemented, that's why prefixed this PR with 'WIP'.

**Pull request merged**